### ISSUE TITLE
Caddyfile global options support for external modules is incomplete

### ISSUE THREAD

**Issue by francislavoie** [caddyserver/caddy#3634]
In an earlier release, I refactored global options in the Caddyfile adapter to provide a `RegisterGlobalOption(name, closure)` function that could be used to register a function to parse arbitrary global options. I thought it would be enough on its own to allow external modules to add their own global options to the Caddyfile, which would in particular be useful for modules that are their own Caddy apps (i.e. @vrongmeal's `git`, @greenpau's `caddy-auth-jwt`, etc.) but it's unfortunately not the case, there's some wiring missing for this to work.

What `RegisterGlobalOption` does is it allows for registering a closure of type `UnmarshalGlobalFunc`, which later gets invoked in `evaluateGlobalOptionsBlock`, that parses the Caddyfile tokens of the given option into an arbitrary `interface{}` (or nil), then it stuffs that struct into an `options` map that gets passed around in `httptype.go`.

Unfortunately, there's no hook anywhere currently to allow external modules to grab their values from `options` to insert their parsed struct at the appropriate places in the JSON config structure. For built-in Caddy apps like `tls`, the parts where its config is pulled out of `options` is hard-coded into `httptype.go`, but external modules don't currently have a mechanism to do the same.

I think we need a new interface that arbitrary Caddy modules can implement which the Caddyfile adapter could invoke to pass along the `options` map so they can pull whatever info they need into the config where appropriate. This part might be a bit tricky, and I'm not sure what the best approach to take here is, so I'm delegating to @mholt 😄

**Comment by mholt** [comment]
Is what you're asking for different from this? https://github.com/caddyserver/caddy/blob/904f149e5b2f0ae9eff09dacb4f1ec41c4a76298/caddyconfig/httpcaddyfile/directives.go#L151-L154

**Comment by francislavoie** [comment]
I think so yes... when would a Caddy app get a hook during the caddyfile adapt phase where they have access to `Helper`? I don't think that's a thing yet.

**Comment by mholt** [comment]
Oh, I see... so a way to delegate/distribute/democratize (something like that?) the construction of the final config structure? I.e. once all the pieces have been put into the pile and it's time to pull them out of the pile and assemble the final config, you want that to be modular?

**Comment by francislavoie** [comment]
Yeah. Something to that effect.

**Comment by francislavoie** [comment]
I'm thinking it should happen right around here?

https://github.com/caddyserver/caddy/blob/master/caddyconfig/httpcaddyfile/httptype.go#L290

**Comment by mholt** [comment]
I think I get it now... but that's really tricky. We _could_ put it there, but by that point, most of the config (anything in `storage` and `apps` keys) is JSON-encoded and stored simply as `json.RawMessage`, meaning that most of the config is inaccessible unless decoded. And you have to know what type to decode into; essentially, you have to load (and possibly provision) the config in order to make it possible to manipulate... either that or transform the raw JSON (eek).

Each piece has to be JSON-encoded from the bottom-up because configs are largely json.RawMessage due to its modular design.

**Comment by francislavoie** [comment]
Maybe caddy apps should get an opportunity to do something like:

```go
cfg.AppsRaw["git"] = caddyconfig.JSON(options["git"], &warnings)
```

Or something like that... :thinking:

**Comment by mholt** [comment]
That only works if you're adding a whole app to the config though...

**Comment by mholt** [comment]
I don't think this needs to be a release blocker for 2.2, so will move this into the 2.x milestone until a good solution emerges.

**Comment by francislavoie** [comment]
Okay, we merged #3990!
